### PR TITLE
fix(spec): stop custom_mask unbounded growth via torch.cat

### LINKED ISSUE #37553: custom_mask grows unbounded via torch.cat in EAGLE/DFlash speculative decoding
### Summary

`EagleVerifyInput.generate_attn_arg_prefill` and `DFlashVerifyInput.generate_attn_arg_prefill` grow the `custom_mask` tensor via `torch.cat` on every batch-size increase and store the result back on `self`. The mask never shrinks, causing unbounded GPU memory growth and stale data returned to the attention kernel.

### Affected files

- `python/sglang/srt/speculative/eagle_info.py:123-136`
- `python/sglang/srt/speculative/dflash_info.py:190-204`

### Root cause

The `custom_mask` tensor is reused across decode steps (stored on `self`). When the batch size increases, `torch.cat` appends `True` padding to make the mask large enough:

```python
self.custom_mask = torch.cat(
    [self.custom_mask, torch.full((mask_numel - self.custom_mask.numel(),), True, ...)],
    dim=0,
)
```

Two problems:

1. **Unbounded memory growth** — `torch.cat` allocates a new tensor every time and the old one becomes garbage. The mask only grows, never shrinks. Over a long-running serving session with fluctuating batch sizes, it accumulates to peak size and stays there, churning the CUDA allocator.

2. **Stale data returned** — when the batch shrinks, the function returns the full oversized `self.custom_mask` (not a slice), so the attention kernel receives more elements than `mask_numel`, including stale `True` padding from a previous larger batch.

The FIXME comments at both sites confirm this was known to be a temporary fix:
```
# FIXME(attn): temporary fix for custom mask padding with cuda graph
```

### Impact

- Slow GPU memory leak in long-running serving workloads with variable batch sizes
- Potential OOM in extreme cases due to allocator fragmentation
- Attention kernel receives stale padding data when batch size decreases

### Environment

- SGLang main (as of 2026-08-30)
- Any GPU, any model using EAGLE or DFlash speculative decoding

### Fix

PR #37120 replaces `torch.cat` with a fresh `torch.full` allocation at exactly `mask_numel` and returns a `[:mask_numel]` slice when the mask is already large enough.

### PR DESCRIPTION
**Pull request by ItzDEXX**
## Purpose
Fixes #37553

`EagleVerifyInput.generate_attn_arg_prefill` (`eagle_info.py`) and `DFlashVerifyInput.generate_attn_arg_prefill` (`dflash_info.py`) grow the `custom_mask` tensor via `torch.cat` on every batch-size increase and store the result back on `self`. The mask never shrinks, so over a long-running serving session with fluctuating batch sizes it:

1. **Leaks GPU memory** — each `torch.cat` allocates a new tensor and abandons the old one. The mask accumulates to peak size and stays there, churning the CUDA allocator.
2. **Returns stale data** — when the batch shrinks, the full oversized mask is returned to the attention kernel, including stale `True` padding from a previous larger batch.

The FIXME comments at both sites confirmed this was a temporary fix.

## Changes

- Replace `torch.cat` with a fresh `torch.full` allocation at exactly `mask_numel` when the current mask is too small.
- Return a `[:mask_numel]` slice when the mask is already large enough (free — just a view, no allocation).
- Remove the FIXME comments.

This stops unbounded growth (the mask is at most the peak `mask_numel`, not the accumulated sum of all increments) and ensures the kernel always receives exactly `mask_numel` elements.

## Why this is safe for CUDA graphs

The CUDA graph captures a kernel that reads from a **capture-time buffer** (`cuda_graph_custom_mask`), not from `self.custom_mask` directly. At every replay step, `call_begin_forward` calls `generate_attn_arg_prefill` to get the mask, then passes it to `wrapper_paged.begin_forward(custom_mask=...)`, which calls FlashInfer's `plan()` — this copies the mask data into the capture-time buffer. The kernel reads from that buffer, so the tensor identity of `self.custom_mask` does not matter.

The AITER backend (`aiter_backend.py:1953-1955`) already copies `spec_info.custom_mask` into its own `cuda_graph_custom_mask` buffer by `spec_info.custom_mask.shape[0]` — a slice has the correct `shape[0]`.

## Test Plan

```bash
python -m pytest test/registered/unit/spec/test_custom_mask_no_leak.py -v
ruff check \
  python/sglang/srt/speculative/eagle_info.py \
  python/sglang/srt/speculative/dflash_info.py \
  test/registered/unit/spec/test_custom_mask_no_leak.py
```

## Test Result

Linux x86_64, Python 3.12, torch 2.13.0+cu132.

```
test/registered/unit/spec/test_custom_mask_no_leak.py    4 passed
  test_eagle_mask_exact_size                    PASSED
  test_eagle_mask_no_unbounded_growth           PASSED
  test_dflash_mask_exact_size                   PASSED
  test_dflash_mask_no_unbounded_growth          PASSED
```

Ruff lint clean on all three files (zero new errors; pre-existing upstream errors in source files are unrelated and out of scope).

## Note for reviewers

The `ngram_info.py` version of this code has the same `torch.cat` pattern but assigns to a local variable (no leak). It does return stale-sized data when the batch shrinks, but that's a separate issue and intentionally left out of this PR to keep the scope focused.

AI assistance was used here. I've reviewed every line, verified the fix against the CUDA graph replay path, and run the tests and lint myself.

### CODE REVIEW

**Comment by ItzDEXX** [comment]
Filed #37553 for tracking. This PR fixes the unbounded `torch.cat` growth described there
